In [1]:
import pandas as pd
data=pd.read_csv('/content/User0_credit_card_transactions.csv')

In [2]:
data.head()

User  Card  Year  Month  Day   Time   Amount           Use Chip  \
0     0     0  2002      9    1  06:21  $134.09  Swipe Transaction   
1     0     0  2002      9    1  06:42   $38.48  Swipe Transaction   
2     0     0  2002      9    2  06:22  $120.34  Swipe Transaction   
3     0     0  2002      9    2  17:45  $128.95  Swipe Transaction   
4     0     0  2002      9    3  06:23  $104.71  Swipe Transaction   

         Merchant Name  Merchant City Merchant State      Zip   MCC Errors?  \
0  3527213246127876953       La Verne             CA  91750.0  5300     NaN   
1  -727612092139916043  Monterey Park             CA  91754.0  5411     NaN   
2  -727612092139916043  Monterey Park             CA  91754.0  5411     NaN   
3  3414527459579106770  Monterey Park             CA  91754.0  5651     NaN   
4  5817218446178736267       La Verne             CA  91750.0  5912     NaN   

  Is Fraud?  
0        No  
1        No  
2        No  
3        No  
4        No

**Example Use in Datasets:**

MCC: 5812 → Eating places and restaurants

MCC: 5411 → Grocery stores and supermarkets

MCC: 5311 → Department stores


In [3]:
data.shape

(19963, 15)

In [4]:
data.describe()

User          Card          Year         Month           Day  \
count  19963.0  19963.000000  19963.000000  19963.000000  19963.000000   
mean       0.0      1.910735   2011.011922      6.568101     15.743876   
std        0.0      1.237763      5.048146      3.477497      8.801378   
min        0.0      0.000000   2002.000000      1.000000      1.000000   
25%        0.0      0.000000   2007.000000      4.000000      8.000000   
50%        0.0      2.000000   2011.000000      7.000000     16.000000   
75%        0.0      3.000000   2015.000000     10.000000     23.000000   
max        0.0      4.000000   2020.000000     12.000000     31.000000   

       Merchant Name           Zip           MCC  
count   1.996300e+04  18316.000000  19963.000000  
mean    7.825653e+17  88812.744922   5617.940239  
std     4.040602e+18  13711.491085    707.982901  
min    -9.179793e+18   1012.000000   1711.000000  
25%    -1.288082e+18  91750.000000   5311.000000  
50%     8.384250e+17  91750.000000   5499.000000  
75%     4.060647e+18  91752.000000   5912.000000  
max     9.137769e+18  99504.000000   9402.000000

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19963 entries, 0 to 19962
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   User            19963 non-null  int64  
 1   Card            19963 non-null  int64  
 2   Year            19963 non-null  int64  
 3   Month           19963 non-null  int64  
 4   Day             19963 non-null  int64  
 5   Time            19963 non-null  object 
 6   Amount          19963 non-null  object 
 7   Use Chip        19963 non-null  object 
 8   Merchant Name   19963 non-null  int64  
 9   Merchant City   19963 non-null  object 
 10  Merchant State  18646 non-null  object 
 11  Zip             18316 non-null  float64
 12  MCC             19963 non-null  int64  
 13  Errors?         574 non-null    object 
 14  Is Fraud?       19963 non-null  object 
dtypes: float64(1), int64(7), object(7)
memory usage: 2.3+ MB


In [6]:
data.isnull().sum()

User                  0
Card                  0
Year                  0
Month                 0
Day                   0
Time                  0
Amount                0
Use Chip              0
Merchant Name         0
Merchant City         0
Merchant State     1317
Zip                1647
MCC                   0
Errors?           19389
Is Fraud?             0
dtype: int64

**Syntax**

dataframe.fillna(value, method, axis, inplace, limit, downcast)

In [7]:
data['Zip'].fillna(data['Zip'].mode()[0], inplace=True)

<ipython-input-7-d8ca4d661ca7>:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Zip'].fillna(data['Zip'].mode()[0], inplace=True)


In [8]:
data['Is Fraud?'] = data['Is Fraud?'].map({'No': 0, 'Yes': 1})

In [9]:
data.head()

User  Card  Year  Month  Day   Time   Amount           Use Chip  \
0     0     0  2002      9    1  06:21  $134.09  Swipe Transaction   
1     0     0  2002      9    1  06:42   $38.48  Swipe Transaction   
2     0     0  2002      9    2  06:22  $120.34  Swipe Transaction   
3     0     0  2002      9    2  17:45  $128.95  Swipe Transaction   
4     0     0  2002      9    3  06:23  $104.71  Swipe Transaction   

         Merchant Name  Merchant City Merchant State      Zip   MCC Errors?  \
0  3527213246127876953       La Verne             CA  91750.0  5300     NaN   
1  -727612092139916043  Monterey Park             CA  91754.0  5411     NaN   
2  -727612092139916043  Monterey Park             CA  91754.0  5411     NaN   
3  3414527459579106770  Monterey Park             CA  91754.0  5651     NaN   
4  5817218446178736267       La Verne             CA  91750.0  5912     NaN   

   Is Fraud?  
0          0  
1          0  
2          0  
3          0  
4          0

In [10]:
data=data.drop('Errors?',axis=1)

In [11]:
data.head()

User  Card  Year  Month  Day   Time   Amount           Use Chip  \
0     0     0  2002      9    1  06:21  $134.09  Swipe Transaction   
1     0     0  2002      9    1  06:42   $38.48  Swipe Transaction   
2     0     0  2002      9    2  06:22  $120.34  Swipe Transaction   
3     0     0  2002      9    2  17:45  $128.95  Swipe Transaction   
4     0     0  2002      9    3  06:23  $104.71  Swipe Transaction   

         Merchant Name  Merchant City Merchant State      Zip   MCC  Is Fraud?  
0  3527213246127876953       La Verne             CA  91750.0  5300          0  
1  -727612092139916043  Monterey Park             CA  91754.0  5411          0  
2  -727612092139916043  Monterey Park             CA  91754.0  5411          0  
3  3414527459579106770  Monterey Park             CA  91754.0  5651          0  
4  5817218446178736267       La Verne             CA  91750.0  5912          0

In [12]:
# Convert Amount column to numeric
data['Amount'] = data['Amount'].replace('[\$,]', '', regex=True).astype(float)

In [16]:
from sklearn.preprocessing import LabelEncoder
categorical_cols = [ 'Use Chip', 'Merchant City', 'Merchant State']
encoder = LabelEncoder()
for col in categorical_cols:
    data[col] = encoder.fit_transform(data[col])

In [23]:
from sklearn.feature_selection import mutual_info_regression
X = data.drop(['Is Fraud?', 'Time'], axis=1)
y=data['Is Fraud?']
mi_scores = mutual_info_regression(X, y)

In [21]:
mi_scores_df = pd.DataFrame({
    'Feature': X.columns,
    'Mutual Information': mi_scores
}).sort_values(by='Mutual Information', ascending=False)

print(mi_scores_df)

           Feature  Mutual Information
11             MCC            0.008525
1             Card            0.002636
10             Zip            0.002526
5           Amount            0.001850
2             Year            0.000863
0             User            0.000000
3            Month            0.000000
4              Day            0.000000
6         Use Chip            0.000000
7    Merchant Name            0.000000
8    Merchant City            0.000000
9   Merchant State            0.000000
